compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirLocalVariableAssignmentAnalyzer.kt

```diff
@@ -0,0 +1,349 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.resolve.dfa
+
+import org.jetbrains.kotlin.contracts.description.EventOccurrencesRange
+import org.jetbrains.kotlin.fir.FirElement
+import org.jetbrains.kotlin.fir.declarations.FirAnonymousFunction
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.declarations.FirProperty
+import org.jetbrains.kotlin.fir.declarations.FirSimpleFunction
+import org.jetbrains.kotlin.fir.expressions.*
+import org.jetbrains.kotlin.fir.references.FirNamedReference
+import org.jetbrains.kotlin.fir.references.FirResolvedNamedReference
+import org.jetbrains.kotlin.fir.resolve.dfa.FirLocalVariableAssignmentAnalyzer.Companion.MiniFlow.Companion.join
+import org.jetbrains.kotlin.fir.visitors.FirVisitor
+import org.jetbrains.kotlin.name.Name
+import org.jetbrains.kotlin.utils.addIfNotNull
+
+/**
+ *  Helper that checks if an access to a local variable access is stable.
+ *
+ *  To determine the stability of an access, call [isAccessToUnstableLocalVariable]. Note that the class contains mutable states. So
+ *  [isAccessToUnstableLocalVariable] only works for an access during the natural FIR tree traversal. This class will not work if one
+ *  queries after the traversal is done.
+ **/
+internal class FirLocalVariableAssignmentAnalyzer(
+    private val assignedLocalVariablesByFunction: Map<FirFunction<*>, AssignedLocalVariables>
+) {
+    /**
+     * Stack storing concurrent lambda arguments for the current visited anonymous function. For example
+     * ```
+     * callWithMultipleLambdaExactlyOnceEach(
+     *   l1 = { x.length },
+     *   l2 = { x = null }
+     * )
+     * ```
+     * From the call, it's indeterministic whether `l1` runs before `l2`. So when handling `l1`, we must mark all variables touched in `l2`
+     * unstable.
+     */
+    private val concurrentLambdaArgsStack: MutableList<Set<FirAnonymousFunction>> = mutableListOf()
+
+    /**
+     * Stack whose element tracks all concurrently modified variables in execution paths other than this one. It's a stack because after
+     * exiting a local function, we must restore the variables to the state before entering this local function.
+     */
+    private val concurrentlyAssignedLocalVariablesStack: MutableList<MutableSet<FirProperty>> = mutableListOf(mutableSetOf())
```

Comment:
Yeah, and it's more than just a never-used sentinel value. When exiting a local function, we need to add all the modified variables to the this default set so that any reads in the root function to a local variable modified inside the lambda can be reported.